File: 0097-interleaving-string/0097-interleaving-string.cpp
class Solution {
public:
    bool isInterleave(string s1, string s2, string s3) {
        if(s1.size()+s2.size() != s3.size()) return false;
        vector<vector<int>> dp(s1.size()+1, vector<int>(s2.size()+1, -1));
        return f(0, 0, s1, s2, s3, dp);
    }
    bool f(int i, int j, string &s1, string &s2, string &s3, vector<vector<int>> &dp){
        if(i+j == s3.size()) return 1;
        if(dp[i][j] != -1) return dp[i][j];
        int ans = 0;
        if(i < s1.size() && s1[i] == s3[i+j]) ans |= f(i+1, j, s1, s2, s3, dp);
        if(j < s2.size() && s2[j] == s3[i+j]) ans |= f(i, j+1, s1, s2, s3, dp);
        return dp[i][j] = ans;
    }
};

File: 0097-interleaving-string/0097-interleaving-string.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        if len(s1)+len(s2) != len(s3): return False
        dp = [[-1]*(len(s2)+1) for _ in range(len(s1)+1)]
        return self.f(0, 0, s1, s2, s3, dp)
    def f(self, i, j, s1, s2, s3, dp):
        if i+j == len(s3): return True
        if dp[i][j] != -1: return dp[i][j]
        ans = 0
        if i < len(s1) and s1[i]==s3[i+j]: ans |= self.f(i+1, j, s1, s2, s3, dp)
        if j < len(s2) and s2[j]==s3[i+j]: ans |= self.f(i, j+1, s1, s2, s3, dp)
        dp[i][j] = ans
        return ans